{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "from pymed import PubMed\n",
    "from metapub import PubMedFetcher \n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "concept1 = '\"precision medicine\"[Title/Abstract] OR \"personalized medicine\"[Title/Abstract] OR \"stratified medicine\"[Title/Abstract] OR \"individualized medicine\"[Title/Abstract] OR \"computational biology\"[Title/Abstract] OR “computational biology”[MeSH Major Topic] OR genom*[Title/Abstract] OR genom*[MeSH Major Topic] OR transcript*[Title/Abstract] OR transcript*[MeSH Major Topic] OR proteom*[Title/Abstract] OR proteom*[MeSH Major Topic] OR lipid*[Title/Abstract] OR lipid*[MeSH Major Topic] OR metagenom*[Title/Abstract] OR metagenom*[MeSH Major Topic] OR exposom*[Title/Abstract] OR exposom*[MeSH Major Topic] OR glycom*[Title/Abstract] OR glycom*[MeSH Major Topic] OR pharmacogenom*[Title/Abstract] OR pharmacogenom*[MeSH Major Topic] OR metabolom*[Title/Abstract] OR metabolom*[MeSH Major Topic] '\n",
    "\n",
    "concept2 = '\"diabetes mellitus, type 2\"[Title/Abstract] OR \"diabetes mellitus, type 2\"[MeSH Major Topic] OR diabetes[Title/Abstract] OR diabetes[MeSH Major Topic] OR glucose[Title/Abstract] OR glucose[MeSH Major Topic] OR diabetic*[Title/Abstract] OR diabetic*[MeSH Major Topic]'\n",
    "\n",
    "years = '(\"2010\"[Date - Publication] : \"2023\"[Date - Publication])'\n",
    "\n",
    "final_query = \"(\" + concept1 + \") AND (\" + concept2 + \") AND \" + years +\")\"\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "https://stackoverflow.com/questions/68952738/get-full-paper-content-from-pubmed-via-api-and-list-of-ids\n",
    "https://pypi.org/project/paperscraper/"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "pubmed = PubMed(tool=\"PubMedSearcher\", email=\"jvargh7@emory.edu\")\n",
    "\n",
    "results = pubmed.query(final_query, max_results=500)\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "'NoneType' object has no attribute 'findall'\n"
     ]
    }
   ],
   "source": [
    "articleList = []\n",
    "articleInfo = []\n",
    "pmid_list = []\n",
    "\n",
    "for article in results:\n",
    "# Print the type of object we've found (can be either PubMedBookArticle or PubMedArticle).\n",
    "# We need to convert it to dictionary with available function\n",
    "    articleDict = article.toDict()\n",
    "    articleList.append(articleDict)\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['Cadambine',\n",
       " 'Expression pattern',\n",
       " 'Methyltransferase',\n",
       " 'Neolamarckia cadamba',\n",
       " 'SABATH gene family']"
      ]
     },
     "execution_count": 26,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "articleList[0]['keywords']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['39526264', '39492480', '39492413', '39492411', '39492355', '39491993', '39491277', '39450278', '39445434', '39372322', '39295610', '39281831', '39280063', '39279783', '39247410', '39183978', '39171278', '39135609', '39076715', '39076452', '39059992', '39055239', '39005872', '39005861', '39003044', '38995299', '38993793', '38966041', '38952883', '38939433', '38938482', '38932907', '38932878', '38932869', '38932848', '38911285', '38904509', '38845737', '38813505', '38813486', '38813008', '38812665', '38799237', '38799007', '38787338', '38784486', '38772847', '38764722', '38736056', '38736053', '38725211', '38725204', '38707926', '38694929', '38680581', '38680576', '38665147', '38647925', '38628008', '38627169', '38619514', '38618550', '38618529', '38618117', '38616871', '38605906', '38602726', '38585559', '38584983', '38577257', '38567068', '38560498', '38550950', '38549489', '38537523', '38527891', '38526828', '38524816', '38523888', '38510479', '38504731', '38487279', '38487001', '38486821', '38481548', '38464790', '38464639', '38464628', '38455935', '38455898', '38455221', '38455204', '38440677', '38440660', '38436919', '38434118', '38433014', '38432863', '38432855', '38431947', '38419721', '38414478', '38414477', '38412665', '38403978', '38391334', '38389952', '38387276', '38387274', '38382279', '38380303', '38379751', '38377720', '38376482', '38371897', '38371896', '38370081', '38370049', '38362587', '38361905', '38357113', '38356679', '38356678', '38352215', '38352208', '38348353', '38348080', '38345591', '38345443', '38344728', '38344688', '38343519', '38333723', '38328582', '38328574', '38328307', '38323222', '38322794', '38322777', '38322643', '38322625', '38322624', '38322325', '38322156', '38320450', '38317711', '38314059', '38313838', '38313837', '38313836', '38313778', '38313077', '38312355', '38310429', '38304905', '38304364', '38304230', '38303689', '38299173', '38299165', '38298928', '38295495', '38293856', '38293554', '38292699', '38292436', '38292115', '38289770', '38289299', '38288476', '38288474', '38288472', '38288471', '38288163', '38288115', '38288074', '38288058', '38283897', '38283757', '38283742', '38283449', '38283334', '38283280', '38282809', '38282661', '38279879', '38279471', '38279421', '38277945', '38275634', '38275633', '38275632', '38274944', '38274229', '38274208', '38274203', '38274107', '38273841', '38272483', '38272481', '38272473', '38269383', '38269247', '38269063', '38266569', '38264894', '38264667', '38264660', '38264652', '38264633', '38264563', '38264328', '38264327', '38264279', '38264278', '38264220', '38264005', '38260808', '38260726', '38260171', '38260162', '38260157', '38260153', '38259870', '38259757', '38259518', '38259490', '38259322', '38259303', '38255639', '38254617', '38251609', '38251552', '38250461', '38250459', '38250336', '38250034', '38249857', '38249612', '38249603', '38248829', '38248828', '38248743', '38247882', '38245457', '38245444', '38245327', '38239990', '38239805', '32778825', '38239732', '38239713', '38239710', '38238958', '38238070', '38236433', '38235430', '38235324', '38235138', '38235001', '38234970', '38234828', '38234627', '38234424', '38233286', '38233270', '38231787', '38231711', '38231706', '38231672', '38230971', '38230286', '38228077', '38227516', '38226278', '38226167', '38226164', '38226160', '38226157', '38225945', '38224715', '38223811', '38223733', '38223708', '38223506', '38223341', '38223220', '38223008', '38223005', '38223004', '38222791', '38222783', '38222188', '38222090', '38220510', '38220509', '38220410', '38220394', '38220343', '38220268', '38215678', '38213890', '38213741', '38213620', '38213389', '38212030', '38212015', '38212002', '38206710', '38206700', '38206683', '38205443', '38205297', '38205242', '38204815', '38204778', '38203703', '38203654', '38203600', '38203573', '38203570', '38203517', '38203514', '38203494', '38203455', '38203359', '38203246', '38203195', '38202777', '38202711', '38202618', '38202201', '38202125', '38202067', '38201924', '38201907', '38201896', '38201855', '38201771', '38201444', '38201252', '38201220', '38201125', '38199135', '38199003', '38198921', '38198415', '38196756', '38196618', '38195352', '38194732', '38193371', '38192830', '38192702', '38192426', '38192280', '38192177', '38190779', '38189542', '38189391', '38189047', '38189046', '38189041', '38188970', '38188526', '38188517', '38188515', '38188508', '38188257', '38187759', '38187751', '38187694', '38187575', '38187496', '38187341', '38187259', '38187250', '38187093', '38186902', '38186860', '38186819', '38186791', '38186569', '38186528', '38186522', '38186506', '38184958', '38184923', '38181524', '38179756', '38179525', '38179455', '38179451', '38179409', '38179344', '38179326', '38179301', '38179020', '38176960', '38176276', '38174672', '38174155', '38174000', '38173904', '38173714', '38173442', '38171805', '38171741', '38171684', '38171202', '38169923', '38169827', '38169733', '38169604', '38169186', '38168419', '38168321', '38165754', '38164418', '38164867', '38164866', '38164832', '38163730', '38163558', '38163557', '38163473', '38163470', '38163138', '38162947', '38162907', '38162681', '38162665', '38162631', '38162524', '38162437', '38162391', '38162055', '38162054', '38162016', '38161967', '38161503', '38161415', '38161391', '38161354', '38161353', '38161338', '38161234', '38161229', '38161128', '38161120', '38161118', '38161102', '38161101', '38160938', '38160917', '38160865', '38160803', '38160800', '38160791', '38160780', '38160746', '38160204', '38159883', '38159882', '38159881', '38159859', '38159812', '38159676', '38159659', '38159629', '38159591', '38159578', '38159023', '38159018', '38159016', '38158979', '38158644', '38158592', '38158337', '38158024', '38157848', '38157803', '38157801', '38157708', '38157301', '38157196', '38157039', '38157037', '38156844', '38156819', '38156400', '38156378', '38156375', '38155901', '38155883', '38155319', '38155161', '38154908', '38154707', '38154610']\n"
     ]
    }
   ],
   "source": [
    "# Generate list of dict records which will hold all article details that could be fetch from PUBMED API\n",
    "for article in articleList:\n",
    "#Sometimes article['pubmed_id'] contains list separated with comma - take first pubmedId in that list - thats article pubmedId\n",
    "    pmid_list.append(article['pubmed_id'].partition('\\n')[0])\n",
    "\n",
    "print(pmid_list)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Plant SABATH methyltransferases are a class of enzymes that catalyze the transfer of the methyl group from S-adenosyl-L-methionine (SAM) to the carboxyl group or the nitrogen group of the substrate to form small molecule methyl esters or N-methylated compounds, which are involved in various secondary metabolite biosynthesis and have important impacts on plant growth, development, and defense reactions. We previously reported the monoterpenoid indole alkaloids (MIAs) cadambine biosynthetic pathway in Neolamarckia cadamba, a woody tree species that provides an important traditional medicine widely used to treat diseases such as diabetes, leprosy, and cancer in Southeast Asia. However, the functions of NcSABATHs in cadambine biosynthesis remain unclear. In this study, 23 NcSABATHs were identified and found to be distributed on 12 of the total 22 chromosomes. Gene structure, conserved motifs, and phylogenetic analysis showed that NcSABATHs could be divided into three groups. According to cis-element analysis, the NcSABATH promoters contained a large number of elements involved in light, plant hormone, and environmental stress responses, as well as binding sites for the BBR-BPC, DOF, and MYB transcription factor families. Based on RNA-seq data and qRT-PCR analysis, the NcSABATH genes exhibited diverse tissue expression patterns. Furthermore, NcSABATH7/22, which clustered with LAMT in the same clade, were both up-regulated under MeJA treatment. The correlation analysis between gene expression and cadambine content showed that NcSABATH7 potentially participated in cadambine biosynthesis. Taken together, our study not only enhanced our understanding of SABATH in N. cadamba but also identified potential candidate genes involved in cadambine biosynthesis.\n",
      "Since its conception some fifty years ago, metabolic control analysis (MCA) aims to understand how cells control their metabolism by adjusting the activity of their enzymes. Here we extend its scope to a whole-cell context. We consider metabolism in the evolutionary context of growth-rate maximisation by optimisation of protein concentrations. This framework allows for the prediction of flux control coefficients from proteomics data or stoichiometric modelling. Since genes compete for finite biosynthetic resources, we treat all protein concentrations as interdependent. We show that elementary flux modes (EFMs) emerge naturally as the optimal metabolic networks in the whole-cell context and we derive their control properties. In the evolutionary optimum, the number of expressed EFMs is determined by the number of protein-concentration constraints that limit growth rate. We use published glucose-limited chemostat data of S. cerevisiae to illustrate that it uses only two EFMs prior to the onset of fermentation and that it uses four EFMs during fermentation. We discuss published enzyme-titration data to show that S. cerevisiae and E. coli indeed can express proteins at growth-rate maximising concentrations. Accordingly, we extend MCA to elementary flux modes operating at an optimal state. We find that the expression of growth-unassociated proteins changes results from classical metabolic control analysis. Finally, we show how flux control coefficients can be estimated from proteomics and ribosome-profiling data. We analyse published proteomics data of E. coli to provide a whole-cell perspective of the control of metabolic enzymes on growth rate. We hope that this paper stimulates a renewed interest in metabolic control analysis, so that it can serve again the purpose it once had: to identify general principles that emerge from the biochemistry of the cell and are conserved across biological species.\n",
      "In order to improve the solubility and hypoglycemic activity of glycyrrhetinic acid (GA), the active mechanism of action of new compounds was explored. A novel 2-(N-3-acetylglycyrrhetinoyl)-N-glucopyranosyl-2-acetamide (compound 9) was synthesized by adding glucosamine (GlcN) to the C-30 carboxyl group of GA, and the hypoglycemic activity mechanism of compound 9 was explored by network pharmacology and molecular docking. The results showed that the solubility of compound 9 was better than GA, and the α-glucosidase inhibitory effect of compound 9 (IC50 = 0.160 mmol/L) was better than GA (IC50 = 0.381 mmol/L). The HepG2 insulin resistance (HepG2-IR) model found that glucose consumption in insulin-resistant cells can increase with the help of GA and compound 9. Network pharmacology screened 268 targets of compound 9 and disease. The core genes in the protein interaction network are epidermal growth factor receptor (EGFR), phosphokinase (SRC), MAPK1, MMP2, mmp9, etc., which are involved in prostate cancer, blood lipid and atherosclerosis, peroxisome proliferation activation receptor (PPAR) signaling pathway, Th17 cell differentiation and other pathways.\n",
      "Sodium-glucose cotransporter 2 (SGLT2, SLC5A2) is a promising target for a new class of drug primarily established as kidney-targeting as well as emerging class of glucose-lowering drugs in diabetes. Studies showed that SGLT2 inhibitors also have a systemic impact via indirectly targeting the heart and kidneys which exerts broad cardio- and nephroprotective effects. Additionally, as cancer cells tightly require glucose supply, studies also questioned how SGLT2 inhibitors impact molecular pathology and cellular metabolism in cancer hallmarks. However, the exact molecular mechanisms responsible for those benefits have not been fully discovered. MicroRNAs (miRNA) and circularRNAs (circRNAs) are endogenous, single-stranded, non-coding RNAs (ncRNAs) that can control protein-coding genes, affecting significant molecular and cellular processes regulating homeostasis. CircRNAs particularly regulate gene expression at the transcriptional and post-transcriptional level by sponging to miRNAs and by altering interactions between proteins.\n",
      "The underlying mechanism associated with the roles of dosed carbon source in denitrification performance remains largely unknown. In this study, three denitrifying consortia (DNC) were constructed via evolutionary top-down enrichment method with well-defined conditions and specific carbon sources (acetate, glucose and their mixture). The reactor operation shows that nearly complete nitrate removal was achieved; however, the glucose feeding resulted in much higher concentrations of biomass and non-settable flocs. The 16S rRNA sequencing suggests that the bacterial diversity of the acetate-fed DNC was significantly higher than those of acetate/glucose-fed and glucose-fed DNCs. The dentrifying population in the acetate-fed DNC was dominated by Propionivibrio (16.1 %) and Thauera (3.4 %); whereas those of acetate/glucose- and glucose-fed DNCs were dominated by Pleomorphomonas (21.5 % and 26.3 %, respectively). Interestingly, the supernatant of acetate-fed DNC contained a high abundance of genera Thauera (averaged at 85.1 %), indicating the free-living nature of Thauera. Both PICURSt2 analysis of 16S rRNA sequencing and metagenomic analysis indicate that the acetate-fed DNC contained higher abundances of denitrifying genes; the acetate/glucose-fed and glucose-fed DNCs, in comparison, enriched genes related to glucose transportation and metabolism. Additionally, the acetate-fed DNC had better network stability than other two groups. This study adds important knowledge regarding the ecological traits of DNC, providing important clues for rational addition of carbon sources in wastewater treatment plants.\n",
      "BACKGROUND: Both air pollution (AP) and impaired lipid metabolism contribute to type 2 diabetes (T2D). However, little is known about the detailed associations of AP to lipidomic markers and the specific lipid metabolomic profile that mediates the impact of AP on incident T2D. We aimed to examine the associations between long-term AP exposure, plasma metabolomic markers, and incident T2D, and subsequently determine the lipid metabolomic profile that mediates the relationship between AP and T2D.\n",
      "METHODS: This prospective study included 82,548 participants from the UK Biobank without a history of T2D at baseline. Baseline plasma samples were analyzed using the nuclear magnetic resonance (NMR) metabolomic platform, which measured 168 metabolomic markers, including lipids, lipoprotein subclasses, and other circulating metabolites. Land Use Regression models were utilized to estimate annual average concentrations of PM<sub>2.5</sub> and NO<sub>2</sub>. The associations among AP, metabolomic markers, and T2D were investigated using multivariable linear regressions and Cox proportional hazards models. Mediation analyses were performed to assess the role of each metabolomic marker in the AP-T2D relationship. Furthermore, principal component (PC) analysis was conducted on 168 metabolomic markers to extract metabolic patterns. These patterns were utilized to determine their associations with AP and T2D, as well as their mediating role in the AP-T2D relationship.\n",
      "RESULTS: We found that long-term AP exposure was associated with some lipid metabolites, including ApoA1, HDL concentration, HDL size, and lipid components within HDL, especially in very large, large, and medium HDL, as well as some other lipids, fatty acids, amino acids, glucose, and glycoprotein acetyls. In pairwise mediation analysis, these metabolites exhibited significant mediation effects in the AP-T2D relationship. We identified six PC<sub>S</sub> representing distinct metabolic patterns. Long-term exposure to PM<sub>2.5</sub> and NO<sub>2</sub> showed significantly negative associations with PC2 (characterized by high levels of ApoA1, larger HDL, other lipids, and low levels of larger VLDL). PC2 mediated 12.3% and 10.3% of the associations of PM<sub>2.5</sub> and NO<sub>2</sub> with incident T2D, respectively.\n",
      "CONCLUSIONS: This study revealed the associations of AP with various lipid metabolites. A lipid metabolomic profile characterized by ApoA1 and larger HDL may mediate the association between AP and incident T2D.\n",
      "20-Hydroxyeicosatetraenoic acid (20-HETE) is a lipid mediator and one of the major arachidonic acid metabolites whose formation is mainly catalyzed by the enzymes cytochrome P450 (CYP) 4F2 and CYP4A11. Several studies have suggested that 20-HETE is involved in the pathogenesis of renal diseases, including diabetic nephropathy and autosomal dominant polycystic kidney disease, and we previously reported compound 1 as a dual inhibitor of CYP4A11/4F2 with therapeutic potential against renal fibrosis. Subsequent studies revealed that compound 1, the dual CYP4A11/4F2 inhibitor, however, exhibited low selectivity over another CYP4F subtype, CYP4F22, which catalyzes ω-hydroxylation of ultra-long-chain fatty acids (ULCFAs); ULCFAs are important for the formation of acylceramides, which play a role in skin barrier formation. Therefore, we sought to develop a CYP4A11/4F2 inhibitor that would show greater CYP4A11/4F2 selectivity against CYP4F22, to avoid potential dermatological side effects. We re-evaluated a series of compounds from our 20-HETE program and identified pyrazolylpyridine derivatives containing a carboxylic acid group showing only weak CYP4F22 inhibition. Subsequent optimization studies from these derivatives led to identification of compound 15, which showed CYP4A11/4F2 inhibition with improved selectivity against CYP4F22. Compound 15 inhibited 20-HETE production in both human and rat renal microsomes and did not inhibit ω-hydroxylation of ULCFAs in human keratinocytes. Compound 15 also significantly inhibited renal 20-HETE production after oral administration.\n",
      "Smoking is a major cause of premature and preventable death. Tobacco exposure has a detrimental effect on many organs and contributes to multiple diseases including chronic obstructive pulmonary disease (COPD), cardiovascular disease, cancer, and diabetes. Cytokines are inflammatory biomarkers that are mechanistically associated with smoking. Machine Learning algorithms allow for the quantitative assessment of the contributions of individual cytokines to tobacco-related diseases. The mapping of cytokines to disease can facilitate and direct treatment modalities. By the application of k Nearest Neighbor (k-NN) and Random Forest machine learning algorithms on 63 plasma cytokines we have demonstrated the classification of smoking. To ensure optimal results, performance improvement techniques such as k-fold cross validation and hyper parameter tuning are employed. Separability efficiency achieved by the models is evaluated using the Area Under the Receiver Operating Characteristic (AUROC) metric. The most significant cytokines that enabled the classification are identified and presented. The statistically significant difference for AUROC score of k-NN and Random Forest has been ascertained using the 2-sample independent t test. A reasonably good classification performance was achieved by k-NN algorithm with an AUROC metric of .87, and a 95% CI of (.823,.917). Random forest exceeded k-NN algorithm's performance, with a perfect AUROC score of 1 and a 95% CI of (1,1). From among the ten most prominent cytokines that contributed to the classification, the ones common to both algorithms are: LIF, IL22, G-CSF/CSF-3, TRAIL. AUROC scores for k-NN and Random Forest are significantly different (p-value = 5.105e-16). The discovery and transference of biomarkers such as cytokines from the platform of molecular investigation to clinical practice, can facilitate precision medicine-based therapeutic interventions.\n",
      "AIMS: To explore the clinical factors and urinary metabolites that predict biopsy-confirmed diabetic nephropathy (DN) in patients with type 2 diabetes mellitus (T2DM).\n",
      "METHODS: Data from the medical records of 126 patients with T2DM who underwent kidney biopsy between January 2010 and October 2020 at a single-center were retrospectively reviewed to investigate the clinical factors that predict DN. Urine samples were collected to perform urine metabolomics in patients with T2DM divided by biopsy-confirmed DN, immunoglobulin A, and membranous nephropathy, and a control group of healthy participants. Each group comprised 11 age- and sex-matched participants. A prediction model was developed using a combination of clinical factors and urinary metabolites, and a multivariate receiver operating characteristic (ROC) analysis was conducted.\n",
      "RESULTS: Age, presence of proliferative diabetic retinopathy, T2DM duration, and hemoglobin A1c levels were clinical factors predictive of DN. Four urinary metabolites (alanine, choline, N-phenylacetylglycine, and trigonelline) had variable importance in projection scores > 1 and were predictive of DN. When conducting multivariate ROC analysis with a combination of clinical factors and urinary metabolites, the area under the curve was 1.000.\n",
      "CONCLUSIONS: The combination of clinical factors and urinary metabolites is highly valuable for predicting biopsy-confirmed DN in patients with T2DM.\n",
      "BACKGROUND: Lifestyle changes in Diabetes Mellitus (DM) positively affect blood glucose and all risk factors. This study aims to determine the effect of swimming and running exercises on oxidant-antioxidant and lipid profiles in streptozotocin (STZ)-induced type 1 diabetic and non-diabetic rats.\n",
      "METHODS: We included forty-eight adult male Wistar albino rats in this study, and we randomly classified them into six groups (eight per group). The groups were organized as Control Sedentary, Control Exercise-swimming, Control Exercise-running (CE-r), Diabetes Sedentary (DS), Diabetes Exercise-swimming (DE-s), and Diabetes Exercise running (DE-r). Half of these rats were subjected to experimental diabetes via STZ. We evaluated total oxidant capacity (TOC), total antioxidant capacity (TAC), superoxide dismutase (SOD), and as lipid parameters: high-density lipoprotein (HDL), low-density lipoprotein (LDL), and triglycerides. The rats were sacrificed at the end of the four weeks.\n",
      "RESULTS: We found a significant difference between DE-s and DE-r groups in terms of TOC (p =0.043) and SOD (p =0.030). The highest TAC was found in the CE-r group, and the highest TOC was found in the DS group. Exercise significantly reduced LDL levels. There was no significant difference between the DE-s and DE-r groups (p =0.084) for lipid profiles (HDL).\n",
      "CONCLUSION: Based on the lower TOC (oxidant) and higher SOD (antioxidant) levels in the diabetic running group, these results suggest that running may be more beneficial than swimming for diabetics. HIPPOKRATIA 2023, 27 (4):148-154.\n",
      "BACKGROUND/AIM: Diabetes mellitus, characterized by hyperglycemia, causes various complications, one of which is memory dysfunction. The frontal lobe is known to be responsible for impaired memory function due to hyperglycemia and is associated with oxidative stress-mediated neuronal cell apoptosis. Chlorogenic acid (CGA) is reported to have neuroprotective effects. However, its effect on the frontal lobe in diabetes mellitus (DM) rats is not widely known. This research aimed to elucidate the effect of CGA on the mRNA expressions of SOD1, SOD2, p53, and Bcl-2 in the frontal lobe of DM rats.\n",
      "MATERIALS AND METHODS: Thirty male Wistar rats (2-month-old, 150-200 gBW) were randomly divided into six groups: C (control), DM1.5 (1.5-month DM), DM2 (2-month DM), CGA12.5, CGA25 and CGA50 (DM+CGA 12.5, 25, and 50 mg/kgBW, respectively). A single dose of streptozotocin (60 mg/kgBW) was intraperitoneally injected. Intraperitoneal CGA injection was administered daily for DM1.5 rats for 14 days. Path length was measured in the Morris water maze (MWM) probe test. After termination, the frontal lobes were carefully harvested for RNA extraction. Reverse transcriptase PCR was performed to examine the mRNA expression of SOD1, SOD2, p53, and Bcl-2.\n",
      "RESULTS: The DM2 group demonstrated significant shorter path length on the MWM probe test and significantly lower mRNA expression of SOD1 and Bcl-2, compared to the C group. After CGA administration, the CGA25 group showed a significantly shorter path length than the C group. The CGA12.5 and CGA25 groups had significantly higher mRNA expression of SOD1 than the DM1.5 group. Compared to the DM1.5 and DM2 groups, SOD2 mRNA expression of the administration of all three CGA doses increased markedly. Furthermore, Bcl-2 mRNA expression was significantly increased in the CGA12.5 and CGA50 groups, compared with the DM2 group.\n",
      "CONCLUSION: Chlorogenic acid might improve memory function through upregulation of frontal lobes' SOD1, SOD2, and Bcl-2 mRNA expression in DM rats.\n",
      "The pathogenic effects of type 2 diabetes on bone tissue are gaining attention, but the cellular and molecular mechanisms underlying osteoimmunology are still unclear in diabetes-related bone diseases. We delineated the single-cell transcriptome of bone marrow cells from both wide type and type 2 diabetes mice, which provided the first detailed global profile of bone marrow cells and revealed a distinct bone immune microenvironment at the genetic level under type 2 diabetic condition. It was observed that osteoclast activity was inhibited due to a dysregulated cytokine network, which ultimately led to decreased osteoclast formation and differentiation. In type 2 diabetes mice, a specific C d 36 + cluster (cluster 18, monocytes/macrophages 2) was identified as the precursor of osteoclasts with diminished differentiation potential. AP-1 was demonstrated to be the key transcription factor in the underlying mechanism.\n",
      "BACKGROUND: Genome-wide association studies for glycemic traits have identified hundreds of loci associated with these biomarkers of glucose homeostasis. Despite this success, the challenge remains to link variant associations to genes, and underlying biological pathways.\n",
      "METHODS: To identify coding variant associations which may pinpoint effector genes at both novel and previously established genome-wide association loci, we performed meta-analyses of exome-array studies for four glycemic traits: glycated hemoglobin (HbA1c, up to 144,060 participants), fasting glucose (FG, up to 129,665 participants), fasting insulin (FI, up to 104,140) and 2hr glucose post-oral glucose challenge (2hGlu, up to 57,878). In addition, we performed network and pathway analyses.\n",
      "RESULTS: Single-variant and gene-based association analyses identified coding variant associations at more than 60 genes, which when combined with other datasets may be useful to nominate effector genes. Network and pathway analyses identified pathways related to insulin secretion, zinc transport and fatty acid metabolism. HbA1c associations were strongly enriched in pathways related to blood cell biology.\n",
      "CONCLUSIONS: Our results provided novel glycemic trait associations and highlighted pathways implicated in glycemic regulation. Exome-array summary statistic results are being made available to the scientific community to enable further discoveries.\n",
      "BACKGROUND: Cardiovascular disease (CVD) is a serious public health problem worldwide. The role of citrus flavanone hesperidin consumption on cardiovascular disease risk factors (CVDRFs) has been examined in many clinical trials, but conflicting results have been found.\n",
      "OBJECTIVES: This study aimed to systematically evaluate the effects of hesperidin extracts or purified hesperidin on CVDRFs in humans with an updated meta-analysis of randomized controlled trials.\n",
      "METHODS: According to the Preferred Reporting Items for Systematic Reviews and Meta-Analyses 2020 guidelines, we systematically screened and searched electronic databases from their establishment to March 2023. Reference lists and previous reviews were also searched. Intervention trials assessing hesperidin consumption on CVD outcomes were included for pooling. To assess the quality of the included articles, the tool of Cochrane risk-of-bias tool was applied. We synthesized the effect sizes with 95% CIs and weighted mean difference (WMD). The I 2 index was used to evaluate the between-study heterogeneity. To explore the heterogeneity source, we used meta-regression and subgroup analysis. Publication bias and sensitivity analysis were also performed. We used the Grading of Recommendations Assessment, Development, and Evaluation approach to evaluate the evidence quality.\n",
      "RESULTS: We included 12 trials with 589 participants. We found evident effects of hesperidin on low-density lipoprotein cholesterol (WMD: -0.22 mmol/L; 95% CI: -0.33, -0.11 mmol/L), total cholesterol (WMD: -0.20 mmol/L; 95% CI: -0.31, -0.08 mmol/L), fasting blood glucose (WMD: -0.15 mg/dL; 95% CI: -0.29, -0.02 mg/dL), quantitative insulin-sensitivity check index (WMD 0.06, 95% CI 0.01 to 0.10), intercellular adhesion molecule 1 (WMD: -13.60 ng/mL; 95% CI: -23.72, -3.48 ng/mL), vascular cell adhesion molecule 1 (WMD: -15.60 ng/mL; 95% CI: -30.13, -1.06 ng/mL), and C-reactive protein (WMD: -0.56 mg/L; 95% CI: -1.11, -0.01 mg/L), whereas no effects were found for other CVDRFs.\n",
      "CONCLUSIONS: Our current findings demonstrate that hesperidin might be advantageous in improving numerous CVDRFs in humans, such as blood lipid concentrations, blood glucose control, and management of inflammatory indicators.\n",
      "Stroke is a major cause of mortality, morbidity and economic burden. Strokes can be thrombotic, embolic or haemorrhagic. The key risk factor for cardioembolic stroke is atrial fibrillation or flutter, and oral anticoagulation (OAC) is recommended in all but the lowest-risk patients with evidence of these arrhythmias. Risk factors for thrombotic stroke overlap strongly with those for other atherosclerotic cardiovascular diseases (ASCVDs). Antiplatelet therapy (APT) should be considered in patients with established ASCVD to reduce risk of cardiovascular events, including stroke. Intensification from single to dual APT or a combination of APT with low-dose OAC can reduce ischaemic stroke risk further, but increases bleeding risk. Blood pressure and lipid profile should be controlled appropriately to guideline targets. In patients with diabetes, good glycaemic control can reduce stroke risk. Inflammation is another emerging target for stroke prevention. Overall, comprehensive assessment and pharmacological modification of risk factors are central to stroke prevention.\n",
      "Maternal diabetes mellitus is among the most frequent environmental contributors to congenital birth defects, including heart defects and craniofacial anomalies, yet the cell types affected and mechanisms of disruption are largely unknown. Using multi-modal single cell analyses, here we show that maternal diabetes affects the epigenomic landscape of specific subsets of cardiac and craniofacial progenitors during embryogenesis. A previously unrecognized cardiac progenitor subpopulation expressing the homeodomain-containing protein ALX3 showed prominent chromatin accessibility changes and acquired a more posterior identity. Similarly, a subpopulation of neural crest-derived cells in the second pharyngeal arch, which contributes to craniofacial structures, displayed abnormalities in the epigenetic landscape and axial patterning defects. Chromatin accessibility changes in both populations were associated with increased retinoic acid signaling, known to establish anterior-posterior identity. This work highlights how an environmental insult can have highly selective epigenomic consequences on discrete cell types leading to developmental patterning defects.\n",
      "Chronic obstructive pulmonary disease (COPD) is a chronic lung disease with limited airflow. COPD is characterized by chronic bronchitis and emphysema, and is often accompanied by malnutrition with fatigue, muscle weakness, and an increased risk of infection. Although the pulmonary function test is used as the gold criterion for diagnosing COPD, it is unable to identify early COPD or classify the subtypes, thereby impeding early intervention and the precise diagnosis of COPD. Recent evidence suggests that metabolic dysfunction, such as changes in lipids, amino acids, glucose, nucleotides, and microbial metabolites in the lungs and intestine, have a great potential for diagnosing COPD in the early stage. However, a comprehensive summary of these metabolites and their effects on COPD is still lacking. This review summarizes the metabolites that are changed in COPD and highlights some promising early diagnostic markers and therapeutic targets. We emphasize that intensified dietary management may be among the most feasible methods to improve metabolism in the body.\n",
      "BACKGROUND: Recent studies suggest that lipid-lowering agents (LLA) may reduce chronic periodontitis, but it is unknown whether this benefit extends to people with type 2 diabetes (T2D).\n",
      "OBJECTIVE: We assessed the association between LLA use and periodontitis in Hispanic adults with T2D.\n",
      "DESIGN: This was a cross-sectional observational study.\n",
      "METHODS: We assessed the association of LLA use and periodontal parameters in 253 Puerto Ricans 40-65 years with T2D who participated in the Lipid-Lowering agents use in Periodontitis and Diabetes Study study. Participants were classified as (a) none- or <1 year, (b) 1-4 years, or (c) >4 years. The primary outcome consists of a tertile percent of sites with probing pocket depth (PPD) ⩾ 4 mm and the secondary outcome includes tertiles of percent sites with clinical attachment loss (CAL) ⩾ 4 mm. Multinomial logistic regression models adjusted for age, gender, smoking status, education, waist circumference, glycosylated hemoglobin A1C (HbA1c), bleeding on probing, examiner, and anti-inflammatory agents were used to estimate the association.\n",
      "RESULTS: LLA (92.5%, statins) was used by 52% of participants. LLA use 1-4 years was associated with lower odds of PPD ⩾ 4 mm (OR: 0.22, p = 0.005; high versus low tertile) or lower odds of CAL ⩾ 4 mm (OR: 0.33, p = 0.02, middle versus low tertile), compared to those with LLA minimal or no use. This association was lost for participants who used LLA for &gt;4 years. LLA users for &gt;4 years with periodontal disease had elevated HbA1c (OR: 1.36, p = 0.05).\n",
      "CONCLUSION: The use of LLA for 1-4 years was associated with lower values of periodontal parameters versus minimal LLA use. This association was not present among people using LLA &gt; 4 years users, but these participants had poorer glycemic control compared to other participants. In this cross-sectional study, the finding that LLA use 1- 4 years is associated with lower values of periodontal parameters of severity in T2D individuals may help clarify some of the controversies regarding the benefit of these medications in this population.\n",
      "BACKGROUND: As a fibrinolytic enzyme from fermented soybean, nattokinase has been shown to be potentially beneficial for cardiovascular health, but current clinical evidences regarding the nattokinase supplementation on cardiovascular risk factors are various. This study aims to evaluate the cardiovascular efficacy of nattokinase.\n",
      "METHODS: Four electronic databases were systematically searched to collect eligible randomized controlled trials. Data were extracted and summarized in a pre-designed form by two independent reviewers. Review Manager 5.4 software (Cochrane Library Software, Oxford, U.K.) was used for meta-analysis and bias risk assessment.\n",
      "RESULTS: Six studies were eligible for quantitative analysis with 546 participants. The overall methodological quality of included studies was high. Relatively low total dosage of nattokinase had a negative effect on blood total cholesterol (MD [mean difference] = 5.27, 95% CI [confidence intervals]: 3.74 to 6.81, p &lt; 0.00001), high-density lipoprotein cholesterol (MD = -2.76, 95% CI: -3.88 to -1.64, p &lt; 0.00001), and low-density lipoprotein cholesterol (MD = 6.49, 95% CI: 0.83 to 12.15, p = 0.02). Nattokinase supplementation significantly reduced systolic blood pressure (MD = -3.45, 95% CI: -4.37 to -2.18, p &lt; 0.00001) and diastolic blood pressure (MD = -2.32, 95% CI: -2.72 to -1.92, p &lt; 0.00001), and led a slight increase in blood glucose (MD = 0.40, 95% CI: 0.20 to 0.60, p &lt; 0.0001) as compared to placebo. Nattokinase group with relatively high total dosage also had a higher total cholesterol (MD = 3.18, 95% CI: 2.29 to 4.06, p &lt; 0.00001) than control interventions, but no significant differences were found in levels of high-density lipoprotein cholesterol and low-density lipoprotein cholesterol. No significant correlation was found between nattokinase supplementation and triglyceride (p = 0.71). No notable adverse events were reported in all studies due to intake of nattokinase.\n",
      "CONCLUSIONS: This study further supports that nattokinase can be used as an effective adjunctive therapy for hypertension, but relatively low-dose supplementation of nattokinase may have no significant lipid-lowering effect. More work will need to be done to determine whether the positive efficacy of nattokinase on cardiovascular risk factors is dose-dependent.\n",
      "SYSTEMATIC REVIEW REGISTRATION: This work has been registered on PROSPERO (CRD42022315020).\n"
     ]
    },
    {
     "ename": "KeyboardInterrupt",
     "evalue": "",
     "output_type": "error",
     "traceback": [
      "\u001b[1;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[1;31mKeyboardInterrupt\u001b[0m                         Traceback (most recent call last)",
      "\u001b[1;32mc:\\code\\external\\equity_precision_llm\\data\\epldat01_retrieving pmids.ipynb Cell 7\u001b[0m line \u001b[0;36m4\n\u001b[0;32m      <a href='vscode-notebook-cell:/c%3A/code/external/equity_precision_llm/data/epldat01_retrieving%20pmids.ipynb#W6sZmlsZQ%3D%3D?line=0'>1</a>\u001b[0m fetch \u001b[39m=\u001b[39m PubMedFetcher()\n\u001b[0;32m      <a href='vscode-notebook-cell:/c%3A/code/external/equity_precision_llm/data/epldat01_retrieving%20pmids.ipynb#W6sZmlsZQ%3D%3D?line=2'>3</a>\u001b[0m \u001b[39mfor\u001b[39;00m pmid \u001b[39min\u001b[39;00m pmid_list:\n\u001b[1;32m----> <a href='vscode-notebook-cell:/c%3A/code/external/equity_precision_llm/data/epldat01_retrieving%20pmids.ipynb#W6sZmlsZQ%3D%3D?line=3'>4</a>\u001b[0m     article \u001b[39m=\u001b[39m fetch\u001b[39m.\u001b[39;49marticle_by_pmid(pmid)\n\u001b[0;32m      <a href='vscode-notebook-cell:/c%3A/code/external/equity_precision_llm/data/epldat01_retrieving%20pmids.ipynb#W6sZmlsZQ%3D%3D?line=4'>5</a>\u001b[0m     \u001b[39mprint\u001b[39m(article\u001b[39m.\u001b[39mabstract)\n",
      "File \u001b[1;32mc:\\Users\\jvargh7\\AppData\\Local\\Programs\\Python\\Python311\\Lib\\site-packages\\metapub\\pubmedfetcher.py:91\u001b[0m, in \u001b[0;36mPubMedFetcher._eutils_article_by_pmid\u001b[1;34m(self, pmid)\u001b[0m\n\u001b[0;32m     89\u001b[0m pmid \u001b[39m=\u001b[39m \u001b[39mstr\u001b[39m(pmid)\n\u001b[0;32m     90\u001b[0m \u001b[39mtry\u001b[39;00m:\n\u001b[1;32m---> 91\u001b[0m     result \u001b[39m=\u001b[39m \u001b[39mself\u001b[39;49m\u001b[39m.\u001b[39;49mqs\u001b[39m.\u001b[39;49mefetch({\u001b[39m'\u001b[39;49m\u001b[39mdb\u001b[39;49m\u001b[39m'\u001b[39;49m: \u001b[39m'\u001b[39;49m\u001b[39mpubmed\u001b[39;49m\u001b[39m'\u001b[39;49m, \u001b[39m'\u001b[39;49m\u001b[39mid\u001b[39;49m\u001b[39m'\u001b[39;49m: pmid})\n\u001b[0;32m     92\u001b[0m \u001b[39mexcept\u001b[39;00m eutils\u001b[39m.\u001b[39mEutilsRequestError:\n\u001b[0;32m     93\u001b[0m     \u001b[39mraise\u001b[39;00m MetaPubError(\u001b[39m'\u001b[39m\u001b[39mInvalid ID \u001b[39m\u001b[39m\"\u001b[39m\u001b[39m%s\u001b[39;00m\u001b[39m\"\u001b[39m\u001b[39m (rejected by Eutils); please check the number and try again.\u001b[39m\u001b[39m'\u001b[39m \u001b[39m%\u001b[39m pmid)\n",
      "File \u001b[1;32mc:\\Users\\jvargh7\\AppData\\Local\\Programs\\Python\\Python311\\Lib\\site-packages\\eutils\\_internal\\queryservice.py:142\u001b[0m, in \u001b[0;36mQueryService.efetch\u001b[1;34m(self, args)\u001b[0m\n\u001b[0;32m    132\u001b[0m \u001b[39mdef\u001b[39;00m \u001b[39mefetch\u001b[39m(\u001b[39mself\u001b[39m, args):\n\u001b[0;32m    133\u001b[0m \u001b[39m    \u001b[39m\u001b[39m\"\"\"\u001b[39;00m\n\u001b[0;32m    134\u001b[0m \u001b[39m    execute a cached, throttled efetch query\u001b[39;00m\n\u001b[0;32m    135\u001b[0m \n\u001b[1;32m   (...)\u001b[0m\n\u001b[0;32m    140\u001b[0m \n\u001b[0;32m    141\u001b[0m \u001b[39m    \"\"\"\u001b[39;00m\n\u001b[1;32m--> 142\u001b[0m     \u001b[39mreturn\u001b[39;00m \u001b[39mself\u001b[39;49m\u001b[39m.\u001b[39;49m_query(\u001b[39m\"\u001b[39;49m\u001b[39m/efetch.fcgi\u001b[39;49m\u001b[39m\"\u001b[39;49m, args)\n",
      "File \u001b[1;32mc:\\Users\\jvargh7\\AppData\\Local\\Programs\\Python\\Python311\\Lib\\site-packages\\eutils\\_internal\\queryservice.py:299\u001b[0m, in \u001b[0;36mQueryService._query\u001b[1;34m(self, path, args, skip_cache, skip_sleep)\u001b[0m\n\u001b[0;32m    296\u001b[0m         _logger\u001b[39m.\u001b[39mdebug(\u001b[39m\"\u001b[39m\u001b[39msleeping \u001b[39m\u001b[39m{sleep_time:.3f}\u001b[39;00m\u001b[39m\"\u001b[39m\u001b[39m.\u001b[39mformat(sleep_time\u001b[39m=\u001b[39msleep_time))\n\u001b[0;32m    297\u001b[0m         time\u001b[39m.\u001b[39msleep(sleep_time)\n\u001b[1;32m--> 299\u001b[0m r \u001b[39m=\u001b[39m requests\u001b[39m.\u001b[39;49mpost(url, full_args)\n\u001b[0;32m    300\u001b[0m \u001b[39mself\u001b[39m\u001b[39m.\u001b[39m_last_request_clock \u001b[39m=\u001b[39m time\u001b[39m.\u001b[39mmonotonic()\n\u001b[0;32m    301\u001b[0m _logger\u001b[39m.\u001b[39mdebug(\u001b[39m\"\u001b[39m\u001b[39mpost(\u001b[39m\u001b[39m{url}\u001b[39;00m\u001b[39m, \u001b[39m\u001b[39m{fas}\u001b[39;00m\u001b[39m): \u001b[39m\u001b[39m{r.status_code}\u001b[39;00m\u001b[39m \u001b[39m\u001b[39m{r.reason}\u001b[39;00m\u001b[39m, \u001b[39m\u001b[39m{len}\u001b[39;00m\u001b[39m)\u001b[39m\u001b[39m\"\u001b[39m\u001b[39m.\u001b[39mformat(\n\u001b[0;32m    302\u001b[0m     url\u001b[39m=\u001b[39murl,\n\u001b[0;32m    303\u001b[0m     fas\u001b[39m=\u001b[39mfull_args_str,\n\u001b[0;32m    304\u001b[0m     r\u001b[39m=\u001b[39mr,\n\u001b[0;32m    305\u001b[0m     \u001b[39mlen\u001b[39m\u001b[39m=\u001b[39m\u001b[39mlen\u001b[39m(r\u001b[39m.\u001b[39mtext)))\n",
      "File \u001b[1;32mc:\\Users\\jvargh7\\AppData\\Local\\Programs\\Python\\Python311\\Lib\\site-packages\\requests\\api.py:115\u001b[0m, in \u001b[0;36mpost\u001b[1;34m(url, data, json, **kwargs)\u001b[0m\n\u001b[0;32m    103\u001b[0m \u001b[39mdef\u001b[39;00m \u001b[39mpost\u001b[39m(url, data\u001b[39m=\u001b[39m\u001b[39mNone\u001b[39;00m, json\u001b[39m=\u001b[39m\u001b[39mNone\u001b[39;00m, \u001b[39m*\u001b[39m\u001b[39m*\u001b[39mkwargs):\n\u001b[0;32m    104\u001b[0m \u001b[39m    \u001b[39m\u001b[39mr\u001b[39m\u001b[39m\"\"\"Sends a POST request.\u001b[39;00m\n\u001b[0;32m    105\u001b[0m \n\u001b[0;32m    106\u001b[0m \u001b[39m    :param url: URL for the new :class:`Request` object.\u001b[39;00m\n\u001b[1;32m   (...)\u001b[0m\n\u001b[0;32m    112\u001b[0m \u001b[39m    :rtype: requests.Response\u001b[39;00m\n\u001b[0;32m    113\u001b[0m \u001b[39m    \"\"\"\u001b[39;00m\n\u001b[1;32m--> 115\u001b[0m     \u001b[39mreturn\u001b[39;00m request(\u001b[39m\"\u001b[39;49m\u001b[39mpost\u001b[39;49m\u001b[39m\"\u001b[39;49m, url, data\u001b[39m=\u001b[39;49mdata, json\u001b[39m=\u001b[39;49mjson, \u001b[39m*\u001b[39;49m\u001b[39m*\u001b[39;49mkwargs)\n",
      "File \u001b[1;32mc:\\Users\\jvargh7\\AppData\\Local\\Programs\\Python\\Python311\\Lib\\site-packages\\requests\\api.py:59\u001b[0m, in \u001b[0;36mrequest\u001b[1;34m(method, url, **kwargs)\u001b[0m\n\u001b[0;32m     55\u001b[0m \u001b[39m# By using the 'with' statement we are sure the session is closed, thus we\u001b[39;00m\n\u001b[0;32m     56\u001b[0m \u001b[39m# avoid leaving sockets open which can trigger a ResourceWarning in some\u001b[39;00m\n\u001b[0;32m     57\u001b[0m \u001b[39m# cases, and look like a memory leak in others.\u001b[39;00m\n\u001b[0;32m     58\u001b[0m \u001b[39mwith\u001b[39;00m sessions\u001b[39m.\u001b[39mSession() \u001b[39mas\u001b[39;00m session:\n\u001b[1;32m---> 59\u001b[0m     \u001b[39mreturn\u001b[39;00m session\u001b[39m.\u001b[39;49mrequest(method\u001b[39m=\u001b[39;49mmethod, url\u001b[39m=\u001b[39;49murl, \u001b[39m*\u001b[39;49m\u001b[39m*\u001b[39;49mkwargs)\n",
      "File \u001b[1;32mc:\\Users\\jvargh7\\AppData\\Local\\Programs\\Python\\Python311\\Lib\\site-packages\\requests\\sessions.py:589\u001b[0m, in \u001b[0;36mSession.request\u001b[1;34m(self, method, url, params, data, headers, cookies, files, auth, timeout, allow_redirects, proxies, hooks, stream, verify, cert, json)\u001b[0m\n\u001b[0;32m    584\u001b[0m send_kwargs \u001b[39m=\u001b[39m {\n\u001b[0;32m    585\u001b[0m     \u001b[39m\"\u001b[39m\u001b[39mtimeout\u001b[39m\u001b[39m\"\u001b[39m: timeout,\n\u001b[0;32m    586\u001b[0m     \u001b[39m\"\u001b[39m\u001b[39mallow_redirects\u001b[39m\u001b[39m\"\u001b[39m: allow_redirects,\n\u001b[0;32m    587\u001b[0m }\n\u001b[0;32m    588\u001b[0m send_kwargs\u001b[39m.\u001b[39mupdate(settings)\n\u001b[1;32m--> 589\u001b[0m resp \u001b[39m=\u001b[39m \u001b[39mself\u001b[39;49m\u001b[39m.\u001b[39;49msend(prep, \u001b[39m*\u001b[39;49m\u001b[39m*\u001b[39;49msend_kwargs)\n\u001b[0;32m    591\u001b[0m \u001b[39mreturn\u001b[39;00m resp\n",
      "File \u001b[1;32mc:\\Users\\jvargh7\\AppData\\Local\\Programs\\Python\\Python311\\Lib\\site-packages\\requests\\sessions.py:703\u001b[0m, in \u001b[0;36mSession.send\u001b[1;34m(self, request, **kwargs)\u001b[0m\n\u001b[0;32m    700\u001b[0m start \u001b[39m=\u001b[39m preferred_clock()\n\u001b[0;32m    702\u001b[0m \u001b[39m# Send the request\u001b[39;00m\n\u001b[1;32m--> 703\u001b[0m r \u001b[39m=\u001b[39m adapter\u001b[39m.\u001b[39;49msend(request, \u001b[39m*\u001b[39;49m\u001b[39m*\u001b[39;49mkwargs)\n\u001b[0;32m    705\u001b[0m \u001b[39m# Total elapsed time of the request (approximately)\u001b[39;00m\n\u001b[0;32m    706\u001b[0m elapsed \u001b[39m=\u001b[39m preferred_clock() \u001b[39m-\u001b[39m start\n",
      "File \u001b[1;32mc:\\Users\\jvargh7\\AppData\\Local\\Programs\\Python\\Python311\\Lib\\site-packages\\requests\\adapters.py:667\u001b[0m, in \u001b[0;36mHTTPAdapter.send\u001b[1;34m(self, request, stream, timeout, verify, cert, proxies)\u001b[0m\n\u001b[0;32m    664\u001b[0m     timeout \u001b[39m=\u001b[39m TimeoutSauce(connect\u001b[39m=\u001b[39mtimeout, read\u001b[39m=\u001b[39mtimeout)\n\u001b[0;32m    666\u001b[0m \u001b[39mtry\u001b[39;00m:\n\u001b[1;32m--> 667\u001b[0m     resp \u001b[39m=\u001b[39m conn\u001b[39m.\u001b[39;49murlopen(\n\u001b[0;32m    668\u001b[0m         method\u001b[39m=\u001b[39;49mrequest\u001b[39m.\u001b[39;49mmethod,\n\u001b[0;32m    669\u001b[0m         url\u001b[39m=\u001b[39;49murl,\n\u001b[0;32m    670\u001b[0m         body\u001b[39m=\u001b[39;49mrequest\u001b[39m.\u001b[39;49mbody,\n\u001b[0;32m    671\u001b[0m         headers\u001b[39m=\u001b[39;49mrequest\u001b[39m.\u001b[39;49mheaders,\n\u001b[0;32m    672\u001b[0m         redirect\u001b[39m=\u001b[39;49m\u001b[39mFalse\u001b[39;49;00m,\n\u001b[0;32m    673\u001b[0m         assert_same_host\u001b[39m=\u001b[39;49m\u001b[39mFalse\u001b[39;49;00m,\n\u001b[0;32m    674\u001b[0m         preload_content\u001b[39m=\u001b[39;49m\u001b[39mFalse\u001b[39;49;00m,\n\u001b[0;32m    675\u001b[0m         decode_content\u001b[39m=\u001b[39;49m\u001b[39mFalse\u001b[39;49;00m,\n\u001b[0;32m    676\u001b[0m         retries\u001b[39m=\u001b[39;49m\u001b[39mself\u001b[39;49m\u001b[39m.\u001b[39;49mmax_retries,\n\u001b[0;32m    677\u001b[0m         timeout\u001b[39m=\u001b[39;49mtimeout,\n\u001b[0;32m    678\u001b[0m         chunked\u001b[39m=\u001b[39;49mchunked,\n\u001b[0;32m    679\u001b[0m     )\n\u001b[0;32m    681\u001b[0m \u001b[39mexcept\u001b[39;00m (ProtocolError, \u001b[39mOSError\u001b[39;00m) \u001b[39mas\u001b[39;00m err:\n\u001b[0;32m    682\u001b[0m     \u001b[39mraise\u001b[39;00m \u001b[39mConnectionError\u001b[39;00m(err, request\u001b[39m=\u001b[39mrequest)\n",
      "File \u001b[1;32mc:\\Users\\jvargh7\\AppData\\Local\\Programs\\Python\\Python311\\Lib\\site-packages\\urllib3\\connectionpool.py:714\u001b[0m, in \u001b[0;36mHTTPConnectionPool.urlopen\u001b[1;34m(self, method, url, body, headers, retries, redirect, assert_same_host, timeout, pool_timeout, release_conn, chunked, body_pos, **response_kw)\u001b[0m\n\u001b[0;32m    711\u001b[0m     \u001b[39mself\u001b[39m\u001b[39m.\u001b[39m_prepare_proxy(conn)\n\u001b[0;32m    713\u001b[0m \u001b[39m# Make the request on the httplib connection object.\u001b[39;00m\n\u001b[1;32m--> 714\u001b[0m httplib_response \u001b[39m=\u001b[39m \u001b[39mself\u001b[39;49m\u001b[39m.\u001b[39;49m_make_request(\n\u001b[0;32m    715\u001b[0m     conn,\n\u001b[0;32m    716\u001b[0m     method,\n\u001b[0;32m    717\u001b[0m     url,\n\u001b[0;32m    718\u001b[0m     timeout\u001b[39m=\u001b[39;49mtimeout_obj,\n\u001b[0;32m    719\u001b[0m     body\u001b[39m=\u001b[39;49mbody,\n\u001b[0;32m    720\u001b[0m     headers\u001b[39m=\u001b[39;49mheaders,\n\u001b[0;32m    721\u001b[0m     chunked\u001b[39m=\u001b[39;49mchunked,\n\u001b[0;32m    722\u001b[0m )\n\u001b[0;32m    724\u001b[0m \u001b[39m# If we're going to release the connection in ``finally:``, then\u001b[39;00m\n\u001b[0;32m    725\u001b[0m \u001b[39m# the response doesn't need to know about the connection. Otherwise\u001b[39;00m\n\u001b[0;32m    726\u001b[0m \u001b[39m# it will also try to release it and we'll have a double-release\u001b[39;00m\n\u001b[0;32m    727\u001b[0m \u001b[39m# mess.\u001b[39;00m\n\u001b[0;32m    728\u001b[0m response_conn \u001b[39m=\u001b[39m conn \u001b[39mif\u001b[39;00m \u001b[39mnot\u001b[39;00m release_conn \u001b[39melse\u001b[39;00m \u001b[39mNone\u001b[39;00m\n",
      "File \u001b[1;32mc:\\Users\\jvargh7\\AppData\\Local\\Programs\\Python\\Python311\\Lib\\site-packages\\urllib3\\connectionpool.py:403\u001b[0m, in \u001b[0;36mHTTPConnectionPool._make_request\u001b[1;34m(self, conn, method, url, timeout, chunked, **httplib_request_kw)\u001b[0m\n\u001b[0;32m    401\u001b[0m \u001b[39m# Trigger any extra validation we need to do.\u001b[39;00m\n\u001b[0;32m    402\u001b[0m \u001b[39mtry\u001b[39;00m:\n\u001b[1;32m--> 403\u001b[0m     \u001b[39mself\u001b[39;49m\u001b[39m.\u001b[39;49m_validate_conn(conn)\n\u001b[0;32m    404\u001b[0m \u001b[39mexcept\u001b[39;00m (SocketTimeout, BaseSSLError) \u001b[39mas\u001b[39;00m e:\n\u001b[0;32m    405\u001b[0m     \u001b[39m# Py2 raises this as a BaseSSLError, Py3 raises it as socket timeout.\u001b[39;00m\n\u001b[0;32m    406\u001b[0m     \u001b[39mself\u001b[39m\u001b[39m.\u001b[39m_raise_timeout(err\u001b[39m=\u001b[39me, url\u001b[39m=\u001b[39murl, timeout_value\u001b[39m=\u001b[39mconn\u001b[39m.\u001b[39mtimeout)\n",
      "File \u001b[1;32mc:\\Users\\jvargh7\\AppData\\Local\\Programs\\Python\\Python311\\Lib\\site-packages\\urllib3\\connectionpool.py:1053\u001b[0m, in \u001b[0;36mHTTPSConnectionPool._validate_conn\u001b[1;34m(self, conn)\u001b[0m\n\u001b[0;32m   1051\u001b[0m \u001b[39m# Force connect early to allow us to validate the connection.\u001b[39;00m\n\u001b[0;32m   1052\u001b[0m \u001b[39mif\u001b[39;00m \u001b[39mnot\u001b[39;00m \u001b[39mgetattr\u001b[39m(conn, \u001b[39m\"\u001b[39m\u001b[39msock\u001b[39m\u001b[39m\"\u001b[39m, \u001b[39mNone\u001b[39;00m):  \u001b[39m# AppEngine might not have  `.sock`\u001b[39;00m\n\u001b[1;32m-> 1053\u001b[0m     conn\u001b[39m.\u001b[39;49mconnect()\n\u001b[0;32m   1055\u001b[0m \u001b[39mif\u001b[39;00m \u001b[39mnot\u001b[39;00m conn\u001b[39m.\u001b[39mis_verified:\n\u001b[0;32m   1056\u001b[0m     warnings\u001b[39m.\u001b[39mwarn(\n\u001b[0;32m   1057\u001b[0m         (\n\u001b[0;32m   1058\u001b[0m             \u001b[39m\"\u001b[39m\u001b[39mUnverified HTTPS request is being made to host \u001b[39m\u001b[39m'\u001b[39m\u001b[39m%s\u001b[39;00m\u001b[39m'\u001b[39m\u001b[39m. \u001b[39m\u001b[39m\"\u001b[39m\n\u001b[1;32m   (...)\u001b[0m\n\u001b[0;32m   1063\u001b[0m         InsecureRequestWarning,\n\u001b[0;32m   1064\u001b[0m     )\n",
      "File \u001b[1;32mc:\\Users\\jvargh7\\AppData\\Local\\Programs\\Python\\Python311\\Lib\\site-packages\\urllib3\\connection.py:419\u001b[0m, in \u001b[0;36mHTTPSConnection.connect\u001b[1;34m(self)\u001b[0m\n\u001b[0;32m    410\u001b[0m \u001b[39mif\u001b[39;00m (\n\u001b[0;32m    411\u001b[0m     \u001b[39mnot\u001b[39;00m \u001b[39mself\u001b[39m\u001b[39m.\u001b[39mca_certs\n\u001b[0;32m    412\u001b[0m     \u001b[39mand\u001b[39;00m \u001b[39mnot\u001b[39;00m \u001b[39mself\u001b[39m\u001b[39m.\u001b[39mca_cert_dir\n\u001b[1;32m   (...)\u001b[0m\n\u001b[0;32m    415\u001b[0m     \u001b[39mand\u001b[39;00m \u001b[39mhasattr\u001b[39m(context, \u001b[39m\"\u001b[39m\u001b[39mload_default_certs\u001b[39m\u001b[39m\"\u001b[39m)\n\u001b[0;32m    416\u001b[0m ):\n\u001b[0;32m    417\u001b[0m     context\u001b[39m.\u001b[39mload_default_certs()\n\u001b[1;32m--> 419\u001b[0m \u001b[39mself\u001b[39m\u001b[39m.\u001b[39msock \u001b[39m=\u001b[39m ssl_wrap_socket(\n\u001b[0;32m    420\u001b[0m     sock\u001b[39m=\u001b[39;49mconn,\n\u001b[0;32m    421\u001b[0m     keyfile\u001b[39m=\u001b[39;49m\u001b[39mself\u001b[39;49m\u001b[39m.\u001b[39;49mkey_file,\n\u001b[0;32m    422\u001b[0m     certfile\u001b[39m=\u001b[39;49m\u001b[39mself\u001b[39;49m\u001b[39m.\u001b[39;49mcert_file,\n\u001b[0;32m    423\u001b[0m     key_password\u001b[39m=\u001b[39;49m\u001b[39mself\u001b[39;49m\u001b[39m.\u001b[39;49mkey_password,\n\u001b[0;32m    424\u001b[0m     ca_certs\u001b[39m=\u001b[39;49m\u001b[39mself\u001b[39;49m\u001b[39m.\u001b[39;49mca_certs,\n\u001b[0;32m    425\u001b[0m     ca_cert_dir\u001b[39m=\u001b[39;49m\u001b[39mself\u001b[39;49m\u001b[39m.\u001b[39;49mca_cert_dir,\n\u001b[0;32m    426\u001b[0m     ca_cert_data\u001b[39m=\u001b[39;49m\u001b[39mself\u001b[39;49m\u001b[39m.\u001b[39;49mca_cert_data,\n\u001b[0;32m    427\u001b[0m     server_hostname\u001b[39m=\u001b[39;49mserver_hostname,\n\u001b[0;32m    428\u001b[0m     ssl_context\u001b[39m=\u001b[39;49mcontext,\n\u001b[0;32m    429\u001b[0m     tls_in_tls\u001b[39m=\u001b[39;49mtls_in_tls,\n\u001b[0;32m    430\u001b[0m )\n\u001b[0;32m    432\u001b[0m \u001b[39m# If we're using all defaults and the connection\u001b[39;00m\n\u001b[0;32m    433\u001b[0m \u001b[39m# is TLSv1 or TLSv1.1 we throw a DeprecationWarning\u001b[39;00m\n\u001b[0;32m    434\u001b[0m \u001b[39m# for the host.\u001b[39;00m\n\u001b[0;32m    435\u001b[0m \u001b[39mif\u001b[39;00m (\n\u001b[0;32m    436\u001b[0m     default_ssl_context\n\u001b[0;32m    437\u001b[0m     \u001b[39mand\u001b[39;00m \u001b[39mself\u001b[39m\u001b[39m.\u001b[39mssl_version \u001b[39mis\u001b[39;00m \u001b[39mNone\u001b[39;00m\n\u001b[0;32m    438\u001b[0m     \u001b[39mand\u001b[39;00m \u001b[39mhasattr\u001b[39m(\u001b[39mself\u001b[39m\u001b[39m.\u001b[39msock, \u001b[39m\"\u001b[39m\u001b[39mversion\u001b[39m\u001b[39m\"\u001b[39m)\n\u001b[0;32m    439\u001b[0m     \u001b[39mand\u001b[39;00m \u001b[39mself\u001b[39m\u001b[39m.\u001b[39msock\u001b[39m.\u001b[39mversion() \u001b[39min\u001b[39;00m {\u001b[39m\"\u001b[39m\u001b[39mTLSv1\u001b[39m\u001b[39m\"\u001b[39m, \u001b[39m\"\u001b[39m\u001b[39mTLSv1.1\u001b[39m\u001b[39m\"\u001b[39m}\n\u001b[0;32m    440\u001b[0m ):\n",
      "File \u001b[1;32mc:\\Users\\jvargh7\\AppData\\Local\\Programs\\Python\\Python311\\Lib\\site-packages\\urllib3\\util\\ssl_.py:449\u001b[0m, in \u001b[0;36mssl_wrap_socket\u001b[1;34m(sock, keyfile, certfile, cert_reqs, ca_certs, server_hostname, ssl_version, ciphers, ssl_context, ca_cert_dir, key_password, ca_cert_data, tls_in_tls)\u001b[0m\n\u001b[0;32m    437\u001b[0m     warnings\u001b[39m.\u001b[39mwarn(\n\u001b[0;32m    438\u001b[0m         \u001b[39m\"\u001b[39m\u001b[39mAn HTTPS request has been made, but the SNI (Server Name \u001b[39m\u001b[39m\"\u001b[39m\n\u001b[0;32m    439\u001b[0m         \u001b[39m\"\u001b[39m\u001b[39mIndication) extension to TLS is not available on this platform. \u001b[39m\u001b[39m\"\u001b[39m\n\u001b[1;32m   (...)\u001b[0m\n\u001b[0;32m    445\u001b[0m         SNIMissingWarning,\n\u001b[0;32m    446\u001b[0m     )\n\u001b[0;32m    448\u001b[0m \u001b[39mif\u001b[39;00m send_sni:\n\u001b[1;32m--> 449\u001b[0m     ssl_sock \u001b[39m=\u001b[39m _ssl_wrap_socket_impl(\n\u001b[0;32m    450\u001b[0m         sock, context, tls_in_tls, server_hostname\u001b[39m=\u001b[39;49mserver_hostname\n\u001b[0;32m    451\u001b[0m     )\n\u001b[0;32m    452\u001b[0m \u001b[39melse\u001b[39;00m:\n\u001b[0;32m    453\u001b[0m     ssl_sock \u001b[39m=\u001b[39m _ssl_wrap_socket_impl(sock, context, tls_in_tls)\n",
      "File \u001b[1;32mc:\\Users\\jvargh7\\AppData\\Local\\Programs\\Python\\Python311\\Lib\\site-packages\\urllib3\\util\\ssl_.py:493\u001b[0m, in \u001b[0;36m_ssl_wrap_socket_impl\u001b[1;34m(sock, ssl_context, tls_in_tls, server_hostname)\u001b[0m\n\u001b[0;32m    490\u001b[0m     \u001b[39mreturn\u001b[39;00m SSLTransport(sock, ssl_context, server_hostname)\n\u001b[0;32m    492\u001b[0m \u001b[39mif\u001b[39;00m server_hostname:\n\u001b[1;32m--> 493\u001b[0m     \u001b[39mreturn\u001b[39;00m ssl_context\u001b[39m.\u001b[39;49mwrap_socket(sock, server_hostname\u001b[39m=\u001b[39;49mserver_hostname)\n\u001b[0;32m    494\u001b[0m \u001b[39melse\u001b[39;00m:\n\u001b[0;32m    495\u001b[0m     \u001b[39mreturn\u001b[39;00m ssl_context\u001b[39m.\u001b[39mwrap_socket(sock)\n",
      "File \u001b[1;32mc:\\Users\\jvargh7\\AppData\\Local\\Programs\\Python\\Python311\\Lib\\ssl.py:517\u001b[0m, in \u001b[0;36mSSLContext.wrap_socket\u001b[1;34m(self, sock, server_side, do_handshake_on_connect, suppress_ragged_eofs, server_hostname, session)\u001b[0m\n\u001b[0;32m    511\u001b[0m \u001b[39mdef\u001b[39;00m \u001b[39mwrap_socket\u001b[39m(\u001b[39mself\u001b[39m, sock, server_side\u001b[39m=\u001b[39m\u001b[39mFalse\u001b[39;00m,\n\u001b[0;32m    512\u001b[0m                 do_handshake_on_connect\u001b[39m=\u001b[39m\u001b[39mTrue\u001b[39;00m,\n\u001b[0;32m    513\u001b[0m                 suppress_ragged_eofs\u001b[39m=\u001b[39m\u001b[39mTrue\u001b[39;00m,\n\u001b[0;32m    514\u001b[0m                 server_hostname\u001b[39m=\u001b[39m\u001b[39mNone\u001b[39;00m, session\u001b[39m=\u001b[39m\u001b[39mNone\u001b[39;00m):\n\u001b[0;32m    515\u001b[0m     \u001b[39m# SSLSocket class handles server_hostname encoding before it calls\u001b[39;00m\n\u001b[0;32m    516\u001b[0m     \u001b[39m# ctx._wrap_socket()\u001b[39;00m\n\u001b[1;32m--> 517\u001b[0m     \u001b[39mreturn\u001b[39;00m \u001b[39mself\u001b[39;49m\u001b[39m.\u001b[39;49msslsocket_class\u001b[39m.\u001b[39;49m_create(\n\u001b[0;32m    518\u001b[0m         sock\u001b[39m=\u001b[39;49msock,\n\u001b[0;32m    519\u001b[0m         server_side\u001b[39m=\u001b[39;49mserver_side,\n\u001b[0;32m    520\u001b[0m         do_handshake_on_connect\u001b[39m=\u001b[39;49mdo_handshake_on_connect,\n\u001b[0;32m    521\u001b[0m         suppress_ragged_eofs\u001b[39m=\u001b[39;49msuppress_ragged_eofs,\n\u001b[0;32m    522\u001b[0m         server_hostname\u001b[39m=\u001b[39;49mserver_hostname,\n\u001b[0;32m    523\u001b[0m         context\u001b[39m=\u001b[39;49m\u001b[39mself\u001b[39;49m,\n\u001b[0;32m    524\u001b[0m         session\u001b[39m=\u001b[39;49msession\n\u001b[0;32m    525\u001b[0m     )\n",
      "File \u001b[1;32mc:\\Users\\jvargh7\\AppData\\Local\\Programs\\Python\\Python311\\Lib\\ssl.py:1075\u001b[0m, in \u001b[0;36mSSLSocket._create\u001b[1;34m(cls, sock, server_side, do_handshake_on_connect, suppress_ragged_eofs, server_hostname, context, session)\u001b[0m\n\u001b[0;32m   1072\u001b[0m         \u001b[39mif\u001b[39;00m timeout \u001b[39m==\u001b[39m \u001b[39m0.0\u001b[39m:\n\u001b[0;32m   1073\u001b[0m             \u001b[39m# non-blocking\u001b[39;00m\n\u001b[0;32m   1074\u001b[0m             \u001b[39mraise\u001b[39;00m \u001b[39mValueError\u001b[39;00m(\u001b[39m\"\u001b[39m\u001b[39mdo_handshake_on_connect should not be specified for non-blocking sockets\u001b[39m\u001b[39m\"\u001b[39m)\n\u001b[1;32m-> 1075\u001b[0m         \u001b[39mself\u001b[39;49m\u001b[39m.\u001b[39;49mdo_handshake()\n\u001b[0;32m   1076\u001b[0m \u001b[39mexcept\u001b[39;00m (\u001b[39mOSError\u001b[39;00m, \u001b[39mValueError\u001b[39;00m):\n\u001b[0;32m   1077\u001b[0m     \u001b[39mself\u001b[39m\u001b[39m.\u001b[39mclose()\n",
      "File \u001b[1;32mc:\\Users\\jvargh7\\AppData\\Local\\Programs\\Python\\Python311\\Lib\\ssl.py:1346\u001b[0m, in \u001b[0;36mSSLSocket.do_handshake\u001b[1;34m(self, block)\u001b[0m\n\u001b[0;32m   1344\u001b[0m     \u001b[39mif\u001b[39;00m timeout \u001b[39m==\u001b[39m \u001b[39m0.0\u001b[39m \u001b[39mand\u001b[39;00m block:\n\u001b[0;32m   1345\u001b[0m         \u001b[39mself\u001b[39m\u001b[39m.\u001b[39msettimeout(\u001b[39mNone\u001b[39;00m)\n\u001b[1;32m-> 1346\u001b[0m     \u001b[39mself\u001b[39m\u001b[39m.\u001b[39m_sslobj\u001b[39m.\u001b[39mdo_handshake()\n\u001b[0;32m   1347\u001b[0m \u001b[39mfinally\u001b[39;00m:\n\u001b[0;32m   1348\u001b[0m     \u001b[39mself\u001b[39m\u001b[39m.\u001b[39msettimeout(timeout)\n",
      "\u001b[1;31mKeyboardInterrupt\u001b[0m: "
     ]
    }
   ],
   "source": [
    "fetch = PubMedFetcher()\n",
    "\n",
    "abstracts_list = []\n",
    "\n",
    "for pmid in pmid_list:\n",
    "    article = fetch.article_by_pmid(pmid)\n",
    "    print(article.abstract)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "10.31083/j.rcm2408234\n"
     ]
    }
   ],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.4"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
